APMrover2/mode.cpp

```diff
@@ -50,56 +49,75 @@ void Mode::set_desired_heading_and_speed(float yaw_angle_cd, float target_speed)
     _desired_speed = target_speed;
 }
 
-void Mode::calc_throttle(float target_speed, bool reversed)
+void Mode::calc_throttle(float target_speed, bool nudge_allowed)
 {
-    // get ground speed from vehicle
-    const float &groundspeed = rover.ground_speed;
+    // add in speed nudging
+    if (nudge_allowed) {
+        target_speed = calc_speed_nudge(target_speed, g.speed_cruise, g.throttle_cruise / 100.0f);
+    }
 
-    // calculate ground speed and ground speed error
-    _speed_error = fabsf(target_speed) - groundspeed;
+    // call throttle controller and convert output to -100 to +100 range
+    float throttle_out = 100.0f * attitude_control.get_throttle_out_speed(target_speed, g2.motors.have_skid_steering(), g2.motors.limit.throttle_lower, g2.motors.limit.throttle_upper, g.speed_cruise, g.throttle_cruise / 100.0f);
 
-    const float throttle_base = (fabsf(target_speed) / g.speed_cruise) * g.throttle_cruise;
-    const float throttle_target = throttle_base + calc_throttle_nudge();
+    // apply limits on throttle
```

Comment:
really?  in any case, this limit checking will be moved into AP_MotorsUGV which is a better place for it in any case.